dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblem.java

```diff
@@ -41,7 +40,7 @@
   private final Symbol symbol;
   private final ClassFile sourceClass;
   private final String symbolProblemMessage;
-  private LinkageProblemCause cause;
+  private final LinkageProblemCause cause;
```

Comment:
Now LinkageProblem is immutable.

Reply:
OK, I see what you'r doing here. However, a setter by another name is still a setter. This would need to eliminate the setCause method, not replace it with withCause, even if that returns a new object. 

In any case, even if we do this sort of refactoring it should be an independnt PR, not mixed in with other changes.

Reply:
When we create LinkageProblem instances, we don't know the causes yet.

Reply:
Then we need is a setter method and it's not an immutable class. Or the problem is we're creating the LinkageProblem at the wrong time.

Reply:
> Or the problem is we're creating the LinkageProblem at the wrong time.

That's an important question. Let's settle that. Do you think the current steps below good/bad?

- Step 1: `LinkageChecker` class finds `LinkageProblem`s from list of `ClassPathEntry`s. This does **not** involve dependency graphs
- Step 2: If a dependency graph is available, LinkageProblemCauseAnnotator annotates the `LinkageProblem`s cause field using the graph.

Reply:
If that's how we do it, then we have a setter method and the object isn't immutable, which might be the way to go. We shouldn't try to hide it. The LinkageProblem object is mutable. 

I would consider whether it's possible to use the dependency graph to annotate the cause at the time the linkage problem is found.